src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,130 @@
+  debounceDelay?: number
+  value?: string
+  onSearch?: (searchTerm: string) => void
+  showSearchButton?: boolean
+}
+
+const shouldInvokeDebounce = (
+  searchTerm: string,
+  showSearchButton?: boolean,
+) => {
+  return Boolean(searchTerm && !showSearchButton)
+}
+
+export function SearchInput({
+  searchIcon = <span>🔍</span>,
+  clearIcon = <span>❎</span>,
+  onSearch,
+  debounceDelay = 300,
+  showSearchButton,
+  value = "",
+  ...inputProps
+}: SearchInputProps) {
+  const [searchTerm, setSearchTerm] = useState("")
+  const [clearButtonHidden, setClearButtonHidden] = useState(true)
+  const timer = useRef<NodeJS.Timeout>()
+
+  useKeyPress(["Enter"], () => {
+    onSearch?.(searchTerm)
+  })
+
+  useEffect(() => {
+    setSearchTerm(value)
+  }, [value])
+
+  useEffect(() => {
+    if (shouldInvokeDebounce(searchTerm, showSearchButton)) {
+      timer.current = setTimeout(() => onSearch?.(searchTerm), debounceDelay)
+    }
+    return () => {
+      clearTimeout(timer.current)
+    }
+  }, [searchTerm, onSearch, showSearchButton, debounceDelay])
+
+  const handleReset = (e: MouseEvent<HTMLButtonElement>) => {
+    e.preventDefault()
+    setSearchTerm("")
+    onSearch?.("")
+    setClearButtonHidden(true)
+  }
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const currentValue = e.currentTarget.value
+    setSearchTerm(currentValue)
+    setClearButtonHidden(!currentValue)
+  }
+
+  const handleBlur = (e: FocusEvent<HTMLInputElement>) => {
+    inputProps.onBlur?.(e)
+    setClearButtonHidden(true)
+  }
+
+  const handleFocus = (e: FocusEvent<HTMLInputElement>) => {
+    inputProps.onFocus?.(e)
+    setClearButtonHidden(!searchTerm.length)
+  }
+
+  const clearButtonClasses = [
+    styles.clearButton,
+    clearButtonHidden && styles.hidden,
+  ].join(" ")
+
+  const inputClasses = [styles.input, inputProps.className].join(" ")
+
+  return (
+    <div className={styles.wrapper}>
+      <div className={styles.container} data-testid="container">
+        <span data-testid="search_icon" className={styles.searchIcon}>
+          {searchIcon}
+        </span>
+        <input
+          {...inputProps}
+          type="search"
+          className={inputClasses}
+          onBlur={handleBlur}
+          onFocus={handleFocus}
+          value={searchTerm}
+          onChange={handleChange}
+        />
+        <button
+          // we are ignoring tabbing since 'esc' is the keyboard accessible means of clearing an input
+          tabIndex={-1}
+          aria-label="click icon to clear search"
+          className={clearButtonClasses}
+          onMouseDown={handleReset}
+        >
+          {clearIcon}
+        </button>
+      </div>
+      {showSearchButton && (
+        <div className={styles.searchButton}>
+          <button
+            aria-label="click to search"
+            onClick={() => searchTerm && onSearch?.(searchTerm)}
+          >
+            Search
+          </button>
+        </div>
+      )}
```

Comment:
This button isn't `debounce`d, meaning that a user could come here and hit search many times and it could trigger many API requests. If this component is handling `debounce` on `onChange` events, would it make sense to also debounce `onClick` events to keep the whole concept consistent? Or does this go beyond the original intent? (Same with the `keyPress` on `Enter`)

Reply:
These are great questions @luiz-chagas. I think you're right. Why wouldn't we just debounce any onSearch event. May not really be necessary to handle the conditional of whether a button is present or not. Let me think on this a bit and maybe rework it. I'll ping you after I have.

Reply:
I think my most recent pass at a debounce handler is the best case scenario. We are now avoiding useEffect and wrapping our debounce function in useMemo to ensure it's not recreated with each render.